mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp

```diff
@@ -488,76 +501,126 @@ coalesceTileLiveRanges(DenseMap<Value, LiveRange> &initialLiveRanges) {
+  if (trivialSpill != allOverlappingRanges.end())
+    return &*trivialSpill;
 
   // Heuristic: Spill the range that ends last (with a compatible tile type).
-  auto isSmallerTileTypeOrEndsEarlier = [](LiveRange *a, LiveRange *b) {
-    return !isTileTypeGreaterOrEqual(a->getTileType(), b->getTileType()) ||
-           a->end() < b->end();
+  auto isSmallerTileTypeOrEndsEarlier = [](LiveRange &a, LiveRange &b) {
+    return !isTileTypeGreaterOrEqual(a.getTileType(), b.getTileType()) ||
+           a.end() < b.end();
   };
-  LiveRange *lastActiveLiveRange = *std::max_element(
-      activeRanges.begin(), activeRanges.end(), isSmallerTileTypeOrEndsEarlier);
-  if (!isSmallerTileTypeOrEndsEarlier(lastActiveLiveRange, newRange))
-    return lastActiveLiveRange;
+  LiveRange &lastActiveLiveRange = *std::max_element(
+      allOverlappingRanges.begin(), allOverlappingRanges.end(),
+      isSmallerTileTypeOrEndsEarlier);
+  if (!isSmallerTileTypeOrEndsEarlier(lastActiveLiveRange, *newRange))
+    return &lastActiveLiveRange;
   return newRange;
 }
 
 /// Greedily allocate tile IDs to live ranges. Spill using simple heuristics.
-/// Note: This does not attempt to fill holes in active live ranges.
 void allocateTilesToLiveRanges(
     ArrayRef<LiveRange *> liveRangesSortedByStartPoint) {
   TileAllocator tileAllocator;
   SetVector<LiveRange *> activeRanges;
+  SetVector<LiveRange *> inactiveRanges;
   for (LiveRange *nextRange : liveRangesSortedByStartPoint) {
-    // Release tile IDs from live ranges that have ended.
     activeRanges.remove_if([&](LiveRange *activeRange) {
+      // Check for live ranges that have expired.
       if (activeRange->end() <= nextRange->start()) {
         tileAllocator.releaseTileId(activeRange->getTileType(),
                                     *activeRange->tileId);
         return true;
       }
+      // Check for live ranges that have become inactive.
+      if (!activeRange->overlaps(nextRange->start())) {
+        tileAllocator.releaseTileId(activeRange->getTileType(),
+                                    *activeRange->tileId);
+        inactiveRanges.insert(activeRange);
+        return true;
+      }
+      return false;
+    });
+    inactiveRanges.remove_if([&](LiveRange *inactiveRange) {
+      // Check for live ranges that have expired.
+      if (inactiveRange->end() <= nextRange->start()) {
+        return true;
+      }
+      // Check for live ranges that have become active.
+      if (inactiveRange->overlaps(nextRange->start())) {
+        tileAllocator.acquireTileId(inactiveRange->getTileType(),
+                                    *inactiveRange->tileId);
+        activeRanges.insert(inactiveRange);
+        return true;
+      }
       return false;
     });
 
+    // Collect inactive live ranges that overlap with the current new live
+    // range. We need to acquire the tile IDs of overlapping inactive ranges to
+    // prevent two (overlapping) live ranges from getting the same tile ID.
```

Comment:
> We need to acquire the tile IDs of overlapping inactive ranges to prevent two (overlapping) live ranges from getting the same tile ID.

I don't follow this sentence. We need to acquire some tile ID:
* if a tile ID is available, use that,
* if no tile ID is available, spill a range (use both active and inactive ranges, but don't prioritise one over the other).

So, IIUC, that sentence should be more like
> We acquire the tile IDs of overlapping **active and inactive** ranges, select one for spilling and re-use it's tile ID for `newRange`.

Is that correct?

Reply:
> Is that correct?

No. This has no relation to spilling or active live ranges.

Reply:
What this does is find all the inactive live ranges that overlap with the and reserves the tile IDs previously assigned to them. As if `newRange` overlaps any `inactiveRange` then it can't fill a hole in the inactive range.

Reply:
This is tested in `holes_in_tile_liveness_inactive_overlaps`.

Reply:
> This has no relation to spilling or active live ranges.

Just to clarify ... L583-L590 find `overlappingInactiveRanges`, which are then indeed used as candidates for spilling on L598. Correct?

Reply:
All live ranges that are overlapping with the current live range as used as spilling candidates, but the point of this code is not just for selecting a tile to spill as your comment suggests. This code always runs (before we know if there's a spill or not), the main point is to reserve tile IDs (to prevent wrongly reusing tile IDs).

Reply:
> This code always runs (before we know if there's a spill or not), the main point is to reserve tile IDs (to prevent wrongly reusing tile IDs).

Sure, but it also happens to find further candidates for spilling, right? So it does two things.

Reply:
The two things it does is:
1. Collect the overlapping inactive live ranges 
2. Reserve the tile IDs for overlapping inactive live ranges

The overlapping ranges are later used for spilling candidates and to release the reserved tile IDs, but neither of those things are happening at this step, so your suggested comment of:
> We acquire the tile IDs of overlapping active and inactive ranges, select one for spilling and re-use it's tile ID for newRange

Would be incorrect and misleading as nothing is being spilled here (and we probably won't spill), and we're also not looking at active ranges.